resolver/resolver.cc

```diff
@@ -453,7 +453,11 @@ class ResolveConstantsWalk {
             }
         } else {
             ENFORCE(resolved.data(ctx)->isClassOrModule());
-            job.klass.data(ctx)->mixins().emplace_back(resolved);
+            auto classData = job.klass.data(ctx);
+            auto fnd = find(classData->mixins().begin(), classData->mixins().end(), resolved);
```

Comment:
`absl::c_find`

Reply:
I expect this fix to be incomplete.
Please try the following example:
```
module F; end;
class A
  include F
end

class B < A
  include F
end
```

Reply:
This program does have problems, but I'm not sure that they're the same issue. If you look at [this sorbet.run example](https://sorbet.run/#%23%20typed%3A%20true%0Amodule%20IFace%0A%20%20extend%20T%3A%3ASig%0A%20%20extend%20T%3A%3AHelpers%0A%20%20interface!%0A%0A%20%20sig%20%7B%20abstract.params(arg%3A%20Integer).void%20%7D%0A%20%20def%20bar(arg)%3B%20end%0Aend%0A%0A%23%20but%20an%20implementation%20cannot%20reasonably%20accept%20a%20subtype%0Aclass%20ArgNarrowing%0A%20%20include%20IFace%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20%7B%20override.params(arg%3A%20String).void%20%7D%0A%20%20def%20bar(arg)%3B%20"%3F"%3B%20end%20%23%20error%3A%20Parameter%20%60arg%60%20of%20type%20%60AbstractVarianceTests%3A%3AC%60%20not%20compatible%20with%20type%20of%20abstract%20method%20%60AbstractVarianceTests%3A%3AIFace%23bar%60%0Aend%0A%0A%23%20but%20an%20implementation%20cannot%20reasonably%20accept%20a%20subtype%0Aclass%20ArgNarrowingTwo%20<%20ArgNarrowing%0A%20%20include%20IFace%0A%20%20extend%20T%3A%3ASig%0A%20%20sig%20%7B%20override.params(arg%3A%20String).void%20%7D%0A%20%20def%20bar(arg)%3B%20"!"%3B%20end%20%23%20error%3A%20Parameter%20%60arg%60%20of%20type%20%60AbstractVarianceTests%3A%3AC%60%20not%20compatible%20with%20type%20of%20abstract%20method%20%60AbstractVarianceTests%3A%3AIFace%23bar%60%0Aend%0A), it'll highlight an error on line 16, but as soon as you edit anything, it'll also highlight an error on line 24, because the errors arising from the child class only appear in incremental. I might be confused here, but it's not apparent to me that that same bug is causing both that issue as well as the one I'm fixing.

Reply:
Sorry, I'm unsure how are two examples related?
I'd expect the example above to be:
1. valid ruby
2. handled correctly in slowpath
3. not handled by this patch(it'll add extra mixin to B where it shouldn't).

Short version: this patch attempts to re-implement part of linearization, and this is not as simple as filtering mixins for duplicates.

Reply:
> This program does have problems, but I'm not sure that they're the same issue.

the snippet above(one with module and two classes, but without methods)
I believe has the same issue: parents will get duplicated.

But this patch won't handle it.

Reply:
My example has the same general structure as your example: one module, two classes (a parent and a child) that both include that same module. By adding methods that have a type signature which doesn't match the module's method, we can induce errors that work as a proxy for how many times the module gets included in that mixin list. If you strip away the methods and sigs and helpers, my example and your example are otherwise the same.

I think maybe the problem here is that I'm not clear on the expected result of linearization: what _should_ `B`'s mixin array be? I had assumed it would contain `F`, but the current behavior is that post-linearization `A` has `F` in its mixin list, while `B` has _nothing_ in its mixin list. (And then the incremental resolver adds an extra copy of `F` to each list, giving `A` two copies and `B` a single copy. This is why an error seems to mysteriously appear after an edit in my above example: because there's no copy of `F` in the mixin list, `B` does not validate its method definition against the module's corresponding method.)

Reply:
Mixins are _additional_ interfaces that were added by a subclass compared to its superclass in order in which they will be selected by virtual dispatch.

In this example, the order of virtual dispatch(and thus linearization) for `B` is `B, A, F`, in that order. Note that `A, F` is the linerization for `A`. 

You can find more tricky examples in https://github.com/sorbet/sorbet/tree/master/test/testdata/resolver/linearization

Reply:
More details in https://en.wikipedia.org/wiki/C3_linearization

Reply:
I think my confusion is that I was coming from an expected typechecking result and not thinking the other way around: it was surprising to me that despite `B`'s explicitly including `F`, we don't actually validate `B`'s method implementations against `F`'s interface definitions, only against its parents. But it does make sense, and no, this PR doesn't fix that issue.